subprojects/core/src/integTest/groovy/org/gradle/internal/tasks/execution/CancellableTaskIntegrationTest.groovy

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.tasks.execution
+
+import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+import org.gradle.integtests.fixtures.daemon.DaemonClientFixture
+import org.gradle.integtests.fixtures.daemon.DaemonIntegrationSpec
+import org.gradle.test.fixtures.ConcurrentTestUtil
+import spock.lang.Unroll
+
+class CancellableTaskIntegrationTest extends DaemonIntegrationSpec implements DirectoryBuildCacheFixture {
+    private static final String START_UP_MESSAGE = "Cancellable task started!"
+    private DaemonClientFixture client
+
+    @Unroll
+    def "custom cancellable task can work when interrupt exception is #scenario"() {
+        given:
+        buildFile << """
+            import org.gradle.api.execution.Cancellable
+
+            class MyCancellableTask extends DefaultTask implements Cancellable {
+                volatile Thread taskExecutionThread
+
+                @TaskAction
+                void run() {
+                    println '$START_UP_MESSAGE'
+                    taskExecutionThread = Thread.currentThread()
+                    try {
+                        new java.util.concurrent.CountDownLatch(1).await()
+                    } catch (InterruptedException e) {
+                        ${scenario == 'swallowed' ? '' : 'throw e'}
```

Comment:
You should always be reseting the thread's interrupted flag when you catch this exception.
`Thread.currentThread().interrupt()`

Reply:
Sure but we can't trust Gradles correctness on people doing that.

Reply:
Ah, I guess I'll have to agree with you there.